packages/synapse-react-client/src/components/ChangePassword/ChangePassword.integration.test.tsx

```diff
@@ -0,0 +1,440 @@
+      expect(changePasswordSpy).toHaveBeenLastCalledWith({
+        concreteType:
+          'org.sagebionetworks.repo.model.auth.ChangePasswordWithTwoFactorAuthToken',
+        newPassword: newPassword,
+        twoFaToken: twoFaToken,
+        userId: userId,
+        otpCode: recoveryCode,
+        otpType: 'RECOVERY_CODE',
+      })
+
+      expect(mockDisplayToast).toHaveBeenCalledWith(
+        'Password successfully changed.',
+        'success',
+      )
+
+      // The form should be reset
+      expect(getCurrentPasswordField()).toHaveValue('')
+      expect(getNewPasswordField()).toHaveValue('')
+      expect(getConfirmPasswordField()).toHaveValue('')
+    })
+  })
+
+  it('shows an error when the 2FA submission is invalid', async () => {
+    const userId = MOCK_USER_ID
+    const twoFaToken = 'mock-2fa-token'
+    const errorMessage = 'The provided code is invalid.'
+    server.use(
+      // The first call will indicate that 2FA is required
+      getRequires2FAChangePasswordHandler(
+        getEndpoint(BackendDestinationEnum.REPO_ENDPOINT),
+        userId,
+        twoFaToken,
+      ),
+      // Update the mock server so the next request will fail with a meaningful error
+      getBadRequestChangePasswordHandler(
+        getEndpoint(BackendDestinationEnum.REPO_ENDPOINT),
+        errorMessage,
+      ),
+    )
+
+    const currentPassword = 'currentPassword'
+    const newPassword = 'newPassword'
+
+    const {
+      user,
+      currentPasswordField,
+      newPasswordField,
+      confirmPasswordField,
+      submitButton,
+    } = setUp()
+
+    await user.type(currentPasswordField, currentPassword)
+    await user.type(newPasswordField, newPassword)
+    await user.type(confirmPasswordField, newPassword)
+
+    await userEvent.click(submitButton)
+
+    // TOTP form should pop up
+    let otpInputs: HTMLElement[] = []
+    await waitFor(() => {
+      otpInputs = screen.getAllByRole('textbox')
+      expect(otpInputs).toHaveLength(6)
+
+      const alert = screen.getByRole('alert')
+      within(alert).getByText(
+        'Two-factor authentication is required to change your password.',
+      )
+
+      expect(mockDisplayToast).not.toHaveBeenCalled()
+      expect(currentPasswordField).not.toBeInTheDocument()
+      expect(newPasswordField).not.toBeInTheDocument()
+      expect(confirmPasswordField).not.toBeInTheDocument()
+
+      expect(changePasswordSpy).toHaveBeenCalledTimes(1)
+      expect(changePasswordSpy).toHaveBeenCalledWith({
+        concreteType:
+          'org.sagebionetworks.repo.model.auth.ChangePasswordWithCurrentPassword',
+        username: mockUserProfileData.userName,
+        newPassword: newPassword,
+        currentPassword: currentPassword,
+      })
+    })
+
+    // Type the code. Once the code is entered, the form should submit automatically
+    for (let i = 0; i < otpInputs.length; i++) {
+      await userEvent.type(otpInputs[i], String(i + 1))
+    }
```

Comment:
Could consider pulling this into a method that can be reused in this test and in the 'supports entering a TOTP if the server requires 2FA' test, but up to you!